File: visionary_welcome/cpp/welcome.cpp
//
// Copyright (c) 2023,2024 SICK AG, Waldkirch
//
// SPDX-License-Identifier: Unlicense

#include <cinttypes>
#include <cstdint>
#include <cstdio>
#include <memory>
#include <string>

#include <iostream>
#include <sstream>

#include <chrono>
#include <thread>

#include <sick_visionary_cpp_base/CoLaParameterWriter.h>
#include <sick_visionary_cpp_base/FrameGrabberBase.h>
#include <sick_visionary_cpp_base/PointCloudPlyWriter.h>
#include <sick_visionary_cpp_base/PointXYZ.h>
#include <sick_visionary_cpp_base/VisionaryControl.h>
#include <sick_visionary_cpp_base/VisionaryType.h>

#include "exitcodes.h"
#include "framewrite.h"
#include "frontendmodes.h"

static ExitCode runWelcomeDemo(visionary::VisionaryType visionaryType, const std::string& ipAddress)
{
  using namespace visionary;
  std::shared_ptr<VisionaryControl> visionaryControl = std::make_shared<VisionaryControl>(visionaryType);

  if (!visionaryControl->open(ipAddress))
  {
    std::fprintf(stderr, "Failed to open control connection to device.\n");

    return ExitCode::eControlCommunicationError;
  }

  // Login to the device for access rights to certain methods
  visionaryControl->login(IAuthentication::UserLevel::SERVICE, "CUST_SERV");
  // Always stop the frontend before using stepAcquisition
  writeFrontendMode(visionaryControl, FrontendMode::eStopped);
  // Logout after configuration
  if (!visionaryControl->logout())
  {
    std::printf("Failed to logout\n");
    return ExitCode::eAuthenticationError;
  }

  // create a frame grabber suitable for the Visionary type used in visionaryControl
  auto pFrameGrabber = visionaryControl->createFrameGrabber();
  // the data handler pointer will later contain the frame data
  auto pDataHandler = visionaryControl->createDataHandler();

  // acquire a single snapshot
  if (!visionaryControl->stepAcquisition())
  {
    std::fprintf(stderr, "Failed to trigger a snapshot\n");

    return ExitCode::eControlCommunicationError;
  }

  // the snapshot has possibly already arrived, so parameter onlyNewer is false
  if (!pFrameGrabber->genGetNextFrame(pDataHandler, false, std::chrono::milliseconds(1200)))
  {
    std::fprintf(stderr, "Frame timeout for snapshot\n");

    return ExitCode::eFrameTimeout;
  }
  else
  {
    std::printf("Frame received in snapshot mode, frame #%" PRIu32 "\n", pDataHandler->getFrameNum());
    // write the frame to disk
    writeFrame(visionaryType, *pDataHandler, "");
    // Convert data to a point cloud
    std::vector<PointXYZ> pointCloud;
    pDataHandler->generatePointCloud(pointCloud);
    pDataHandler->transformPointCloud(pointCloud);
    // Write point cloud to PLY
    const std::string framePrefix  = std::to_string(pDataHandler->getFrameNum());
    std::string       plyFilePath  = framePrefix + "-pointcloud.ply";
    const char*       cPlyFilePath = plyFilePath.c_str();
    std::printf("Writing frame to %s\n", cPlyFilePath);
    if (visionaryType == VisionaryType::eVisionaryS)
      PointCloudPlyWriter::WriteFormatPLY(cPlyFilePath, pointCloud, pDataHandler->getRGBAMap(), true);
    else
      PointCloudPlyWriter::WriteFormatPLY(cPlyFilePath, pointCloud, pDataHandler->getIntensityMap(), true);
    std::printf("Finished writing frame to %s\n", cPlyFilePath);
  }

  // Another login to change frontend parameters
  if (!visionaryControl->login(IAuthentication::UserLevel::AUTHORIZED_CLIENT, "CLIENT"))
  {
    std::printf("Failed to log into the device.\n");
    return ExitCode::eAuthenticationError;
  }
  if (visionaryType.toString() == "Visionary-S")
  {
    // set_integrationTime
    CoLaCommand setIntegrationTimeUsCommand =
      CoLaParameterWriter(CoLaCommandType::WRITE_VARIABLE, "integrationTimeUs").parameterUDInt(3000).build();
    auto setIntegrationTimeUsResponse = visionaryControl->sendCommand(setIntegrationTimeUsCommand);
    std::cout << "Set integration time to 3000 micro seconds." << "\n";
  }
  else if (visionaryType.toString() == "Visionary-T_Mini")
  {
    // Set FramePeriodUS
    CoLaCommand setFramePeriodUsCommand =
      CoLaParameterWriter(CoLaCommandType::WRITE_VARIABLE, "framePeriodUs").parameterUDInt(60000).build();
    auto setFramePeriodUsResponse = visionaryControl->sendCommand(setFramePeriodUsCommand);
    std::cout << "Set frame period to 60000 micro seconds." << "\n";
  }

  if (!visionaryControl->logout())
  {
    std::printf("Failed to logout\n");
    return ExitCode::eAuthenticationError;
  }

  // acquire a single snapshot
  if (!visionaryControl->stepAcquisition())
  {
    std::fprintf(stderr, "Failed to trigger a snapshot\n");

    return ExitCode::eControlCommunicationError;
  }

  // the snapshot has possibly already arrived, so parameter onlyNewer is false
  if (!pFrameGrabber->genGetNextFrame(pDataHandler, false, std::chrono::milliseconds(2000)))
  {
    std::fprintf(stderr, "Frame timeout for snapshot\n");

    return ExitCode::eFrameTimeout;
  }
  else
  {
    std::printf("Frame received in snapshot mode, frame #%" PRIu32 "\n", pDataHandler->getFrameNum());
    // write the frame to disk
    writeFrame(visionaryType, *pDataHandler, "");
    // Convert data to a point cloud
    std::vector<PointXYZ> pointCloud;
    pDataHandler->generatePointCloud(pointCloud);
    pDataHandler->transformPointCloud(pointCloud);
    // Write point cloud to PLY
    const std::string framePrefix  = std::to_string(pDataHandler->getFrameNum());
    std::string       plyFilePath  = framePrefix + "-pointcloud.ply";
    const char*       cPlyFilePath = plyFilePath.c_str();
    std::printf("Writing frame to %s\n", cPlyFilePath);
    if (visionaryType == VisionaryType::eVisionaryS)
      PointCloudPlyWriter::WriteFormatPLY(cPlyFilePath, pointCloud, pDataHandler->getRGBAMap(), true);
    else
      PointCloudPlyWriter::WriteFormatPLY(cPlyFilePath, pointCloud, pDataHandler->getIntensityMap(), true);
    std::printf("Finished writing frame to %s\n", cPlyFilePath);
  }

  // Before exiting the sample we reset the frontendmode to "Continuous" and delete the framegrabber
  if (!visionaryControl->login(IAuthentication::UserLevel::AUTHORIZED_CLIENT, "CLIENT"))
  {
    std::printf("Failed to log into the device.\n");
    return ExitCode::eAuthenticationError;
  }
  
  writeFrontendMode(visionaryControl, FrontendMode::eContinuous);
  // delete the frame grabber
  pFrameGrabber.reset();

  visionaryControl->logout();
  visionaryControl->close();
  std::printf("Logout and close.\n");

  return ExitCode::eOk;
}

int main(int argc, char* argv[])
{
  std::string deviceIpAddr{"192.168.1.10"};
  std::string filePrefix{""};

  visionary::VisionaryType visionaryType(visionary::VisionaryType::eVisionaryTMini);

  bool showHelpAndExit = false;

  ExitCode exitCode = ExitCode::eOk;

  for (int i = 1; i < argc; ++i)
  {
    std::istringstream argstream(argv[i]);

    if (argstream.get() != '-')
    {
      showHelpAndExit = true;
      exitCode        = ExitCode::eParamError;
      break;
    }
    switch (argstream.get())
    {
      case 'h':
        showHelpAndExit = true;
        break;
      case 'i':
        argstream >> deviceIpAddr;
        break;
      case 'd':
      {
        std::string visionaryTypeName;
        argstream >> visionaryTypeName;
        try
        {
          visionaryType = visionary::VisionaryType::fromString(visionaryTypeName);
        }
        catch (const std::invalid_argument& e)
        {
          // NOLINTNEXTLINE(performance-avoid-endl)
          std::cerr << e.what() << ": '" << visionaryTypeName << "'" << std::endl;
          showHelpAndExit = true;
          exitCode        = ExitCode::eParamError;
        }
        break;
      }
      default:
        showHelpAndExit = true;
        break;
    }
  }

  if (showHelpAndExit)
  {
    std::cout << "\nUsage: " << argv[0] << " [option]*\n";

    std::cout << "where option is one of\n";
    std::cout << "-h              show this help and exit\n";
    std::cout << "-i<IP>          connect to the device with IP address <IP>; default is " << deviceIpAddr << '\n';
    std::cout << "-d<device type> visionary product type; default is '" << visionaryType.toString() << "'\n";

    std::cout << "\nVisionary product types:\n";
    for (const auto& name : visionary::VisionaryType::getNames())
    {
      std::cout << "  " << name << '\n';
    }

    return static_cast<int>(exitCode);
  }

  exitCode = runWelcomeDemo(visionaryType, deviceIpAddr);

  std::cout << "exit code " << static_cast<int>(exitCode) << '\n';

  return static_cast<int>(exitCode);
}

File: visionary_welcome/python/welcome.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Copyright (c) 2024 SICK AG, Waldkirch
#
# SPDX-License-Identifier: Unlicense

import argparse
import os
import time
from time import sleep

from base.python.Control import Control
from base.python.PointCloud.PointCloud import (convertToPointCloud,
                                               convertToPointCloudOptimized,
                                               writePointCloudToPCD,
                                               writePointCloudToPLY)
from base.python.Stream import Streaming
from base.python.Streaming import Data
from base.python.Streaming.BlobServerConfiguration import BlobClientConfig
from shared.python.framewrite import writeFrame
from shared.python.devices_config import get_device_config
from base.python.Usertypes import FrontendMode


def runWelcomeDemo(ip_address: str, cola_protocol: str, control_port: int, device_type: str):
    # Directories to save the output in
    pcl_dir = 'VisionaryToPointCloud'
    img_dir = 'VisionaryImages'
    os.makedirs(pcl_dir, exist_ok=True)
    os.makedirs(img_dir, exist_ok=True)

    # Connect to control channel
    device_control = Control(ip_address, cola_protocol, control_port)
    device_control.open()

    # Login to the device for access rights to certain methods
    device_control.login(Control.USERLEVEL_SERVICE, 'CUST_SERV')

    # Streaming settings
    streaming_settings = BlobClientConfig(device_control)
    streaming_device = None
    streaming_settings.setTransportProtocol(
        streaming_settings.PROTOCOL_TCP)
    streaming_settings.setBlobTcpPort(2114)
    # Open streaming channel
    streaming_device = Streaming(ip_address, 2114)
    streaming_device.openStream()

    # Stop image acquisition (works always, also when already stopped)
    # Further you should always stop the device before calling singleStep()
    # tag::stop_image_acquisition[]
    device_control.setFrontendMode(FrontendMode.Stopped)
    # end::stop_image_acquisition[]

    # Logout after configuration
    device_control.logout()

    # Take a snapshot
    device_control.singleStep()
    sensor_data = Data.Data()
    streaming_device.getFrame()
    whole_frame = streaming_device.frame
    sensor_data.read(whole_frame, convertToMM=False)
    print("Data Timestamp [YYYY-MM-DD HH:MM:SS.mm] = %04u-%02u-%02u %02u:%02u:%02u.%03u" % (
        sensor_data.getDecodedTimestamp()))
    if sensor_data.hasDepthMap:
        frame_number = sensor_data.depthmap.frameNumber
        print("Data contains depth map data:")
        print("=== Write PNG file: Frame number: {}".format(frame_number))
        writeFrame(device_type, sensor_data,
                   os.path.join(img_dir, ""))
        print("=== Converting image to pointcloud")

        # Non optimized
        start_time = time.time()
        world_coordinates, dist_data = convertToPointCloud(sensor_data.depthmap.distance,
                                                           sensor_data.depthmap.intensity,
                                                           sensor_data.depthmap.confidence,
                                                           sensor_data.cameraParams, sensor_data.xmlParser.stereo)
        end_time = time.time()
        execution_time = end_time - start_time
        print(f"convertToPointCloud took: {execution_time:.3}s")

        # Optimized
        is_stereo = True if device_type == "Visionary-S" else False
        start_time = time.time()
        point_cloud = convertToPointCloudOptimized(sensor_data.depthmap.distance,
                                                   sensor_data.depthmap.confidence,
                                                   sensor_data.cameraParams, is_stereo)
        end_time = time.time()
        execution_time = end_time - start_time
        print(f"convertToPointCloudOptimized took: {execution_time:.3}s")

        # Write output of the non optimized function to PLY
        writePointCloudToPLY(os.path.join(
            pcl_dir, "world_coordinates{}.ply".format(frame_number)), world_coordinates)

        # Write output of the optimized function to PCD
        writePointCloudToPCD(os.path.join(
            pcl_dir, "world_coordinates{}.pcd".format(frame_number)), point_cloud.reshape(-1, point_cloud.shape[-1]))

    # Login to the device for access rights to certain methods
    device_control.login(Control.USERLEVEL_SERVICE, 'CUST_SERV')

    # change one frontend setting and take another image
    if device_type == "Visionary-T Mini":
        device_control.setFramePeriodUs(60000)
        print("\nSet frame period to 60000 micro seconds.\n")
    elif device_type == "Visionary-S":
        device_control.setIntegrationTimeUs(3000)
        print("\nSet integration time to 3000 micor seconds.\n")

    # Logout after configuration
    device_control.logout()

    # Take a snapshot
    device_control.singleStep()
    sensor_data = Data.Data()
    streaming_device.getFrame()
    whole_frame = streaming_device.frame
    sensor_data.read(whole_frame, convertToMM=False)
    print("Data Timestamp [YYYY-MM-DD HH:MM:SS.mm] = %04u-%02u-%02u %02u:%02u:%02u.%03u" % (
        sensor_data.getDecodedTimestamp()))
    if sensor_data.hasDepthMap:
        frame_number = sensor_data.depthmap.frameNumber
        print("Data contains depth map data:")
        print("=== Write PNG file: Frame number: {}".format(frame_number))
        writeFrame(device_type, sensor_data,
                   os.path.join(img_dir, ""))
        print("=== Converting image to pointcloud")

        # Non optimized
        start_time = time.time()
        world_coordinates, dist_data = convertToPointCloud(sensor_data.depthmap.distance,
                                                           sensor_data.depthmap.intensity,
                                                           sensor_data.depthmap.confidence,
                                                           sensor_data.cameraParams, sensor_data.xmlParser.stereo)
        end_time = time.time()
        execution_time = end_time - start_time
        print(f"convertToPointCloud took: {execution_time:.3}s")

        # Optimized
        is_stereo = True if device_type == "Visionary-S" else False
        start_time = time.time()
        point_cloud = convertToPointCloudOptimized(sensor_data.depthmap.distance,
                                                   sensor_data.depthmap.confidence,
                                                   sensor_data.cameraParams, is_stereo)
        end_time = time.time()
        execution_time = end_time - start_time
        print(f"convertToPointCloudOptimized took: {execution_time:.3}s")

        # Write output of the non optimized function to PLY
        writePointCloudToPLY(os.path.join(
            pcl_dir, "world_coordinates{}.ply".format(frame_number)), world_coordinates)

        # Write output of the optimized function to PCD
        writePointCloudToPCD(os.path.join(
            pcl_dir, "world_coordinates{}.pcd".format(frame_number)), point_cloud.reshape(-1, point_cloud.shape[-1]))

    # Close streaming socket
    streaming_device.closeStream()
    device_control.login(Control.USERLEVEL_AUTH_CLIENT, 'CLIENT')
    # Reset the image acquisition to default mode
    device_control.setFrontendMode(FrontendMode.Continuous)
    device_control.logout()
    device_control.close()
    print("Logout and close.")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="This welcome demo shows what is possible with the API.")
    parser.add_argument('-i', '--ipAddress', required=False, type=str,
                        default="192.168.1.10", help="The ip address of the device.")
    parser.add_argument('-d', '--device_type', required=False, type=str,
                        default="Visionary-T Mini", choices=["Visionary-S", "Visionary-T Mini"],
                        help="Visionary product type.")
    args = parser.parse_args()

    cola_protocol, control_port, _ = get_device_config(args.device_type)

    runWelcomeDemo(args.ipAddress, cola_protocol,
                   control_port, args.device_type)
